site/guides/effective-go-krm-function.md

```diff
@@ -0,0 +1,212 @@
+KRM resources from the input in `items` fields.
+```go
+func (r *FunctionX) Run(ctx *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+      // TODO: Add your KRM resource mutate or validate logic.
+}
+
+func main() {
+    if err := fn.AsMain(&FunctionX{}); err != nil {
+        os.Exit(1)
+    }
+}
+```
+
+<!--- TODO: we should not require users to understand and provide the input ResourceList. 
+We should build a test infra that users only provide the KRM resources--->
+
+### Select KRM resources 
+The `fn.KubeObjects` is a slice of `*fn.KubeObject`, that you can apply some select logic to easily choose
+the target KRM resources. See below example on using `Where` and `WhereNot` to filter different types of resources.
+```go
+func (r *YourFunction) Run(context *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+	// namespaceScoped contains only namespace scoped resources  
+    namespaceScoped := objects.Where(func(o *fn.KubeObject) bool { return o.IsNamespaceScoped() }) 
+    // clusterScoped contains only cluster scoped resources
+    clusterScoped := objects.Where(func(o *fn.KubeObject) bool { return o.IsClusterScoped() })
+    // customDeployment contains all resources of Kind "CustomDeployment", in Group "fn.kpt.dev" with any Versions.  
+    customDeployment := objects.Where(fn.IsGVK("fn.kpt.dev", "", "CustomDeployment") })
+    // excluded contains all resources except namespace objects
+    excluded := objects.WhereNot(fn.IsGVK("v1", "", "Namespace") })
+}
+```
+
+### Read and write a field spec path of a KRM resource
+
+Like [unstructured.Unstructured], `fn.KubeObject` (and `fn.SubObject`) provides a series of methods to 
+let you read and write different resources types.
+
+```go
+func (r *YourFunction) Run(context *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+	// Get first deployment object.
+	deployment := items.Where(fn.IsGVK("apps", "v1", "Deployment")).Where(func(o *fn.KubeObject) bool{return o.GetName() == "nginx"})[0]
+	// Get the int value from deployment `spec.replicas`
+	replicas := deployment.NestedInt64OrDie("spec", "replicas")
+	fn.Logf("replicas is %v\n", replicas)
+    // Get the boolean value from deployment `spec.paused`
+    paused := deployment.NestedBoolOrDie("spec", "paused")
+    fn.Logf("paused is %v\n", paused)
+    // Update strategy from Recreate to RollingUpdate.
+    if strategy := obj.NestedStringOrDie("spec", "strategy", "type"); strategy == "Recreate" {
+    	obj.SetNestedStringOrDie("RollingUpdate", "spec", "strategy", "type")
+    }
+}
+```
+
+Besides the [unstructured.Unstructured] style, you can also run functions on each sub-field as a `fn.SubObject`  
+```go
+func (r *YourFunction) Run(context *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+    // Get first deployment object.
+	deployment := items.Where(fn.IsGVK("apps", "v1", "Deployment")).Where(func(o *fn.KubeObject) bool{return o.GetName() == "nginx"})[0]
```

Comment:
I think it's okay to have this expression in multiple lines that helps with readability.